### PR TITLE
Add presense to puzzle page

### PR DESCRIPTION
**Pull request by digit01Wave**
#1031

Add Presence on puzzle list

At two players
image

At three players
image

### CODE REVIEW

File: ServerCore/Pages/Components/PresenceComponent.razor.cs [hunk 1]
```diff
@@ -33,6 +33,9 @@ public partial class PresenceComponent : IAsyncDisposable
         [Parameter]
         public int PuzzleId { get; set; }
 
+        [Parameter]
+        public bool IsReadOnly { get; set; }
```
Comment (tabascq) [inline ServerCore/Pages/Components/PresenceComponent.razor.cs:37]: I think this is a somewhat weird name for the property. Wondering if "IsListComponent" is clearer.

**Review by tabascq — COMMENTED** [review]
I think this is fantastic (modulo the one very minor comment I made), but I want us to think VERY carefully about the fact that we are having scale troubles and whether this change will make us more likely to run into performance issues day-of. @morganbr needs to be in that conversation.

**Comment by morganbr** [comment]
> I think this is fantastic (modulo the one very minor comment I made), but I want us to think VERY carefully about the fact that we are having scale troubles and whether this change will make us more likely to run into performance issues day-of. @morganbr needs to be in that conversation.

If this doesn't create extra Blazor connections, which I think it doesn't and we can confirm, the perf should be ok.

File: ServerCore/Pages/Components/PresenceComponent.razor.cs (hunk 1 above)
Comment (morganbr) [reply ServerCore/Pages/Components/PresenceComponent.razor.cs:37]: Maybe add a comment for this? ReadOnly is a good concept for this (that it shows presence, but doesn't write it), but I didn't immediately get it either

File: ServerCore/Pages/Components/PresenceComponent.razor.cs [hunk 2]
```diff
@@ -106,7 +109,11 @@ protected override async Task OnParametersSetAsync()
 
             await UpdateModelAsync(teamPuzzleStore.PresentPages);
 
-            await MessageHub.BroadcastPresenceMessageAsync(new PresenceMessage { PageInstance = pageInstance, PuzzleUserId = PuzzleUserId, TeamId = TeamId, PuzzleId = PuzzleId, PresenceType = PresenceType.Active });
+            if (!this.IsReadOnly)
```
Comment (morganbr) [inline ServerCore/Pages/Components/PresenceComponent.razor.cs:112]: This check is also needed in DisposeAsync where it shows the player disconnecting from the puzzle

File: ServerCore/Pages/Puzzles/Play.cshtml [hunk 3]
```diff
@@ -178,6 +184,12 @@ else
                             }
                         </td>
                     }
+                    else
+                    {
+                        <td>
+                            <component type="typeof(ServerCore.Pages.Components.PresenceComponent)" render-mode="Server" param-PuzzleUserId="@Model.LoggedInUser.ID" param-TeamId="@Model.Team.ID" param-PuzzleId="@item.ID" param-IsReadOnly=true />
```
Comment (morganbr) [inline ServerCore/Pages/Puzzles/Play.cshtml:190]: The full list of names might get crowded on the puzzle list if several players join a puzzle. Consider shortening it. Maybe something like for 3+ players, show a count?

**Review by morganbr — APPROVED** [review]
Thanks for doing this! At the meeting, let's confirm the effect on Blazor/SignalR connections

File: ServerCore/Pages/Puzzles/Play.cshtml (hunk 3 above)
Comment (morganbr) [inline ServerCore/Pages/Puzzles/Play.cshtml:190]: This needs to check Event.AllowBlazor

**Review by morganbr — APPROVED** [review]

**Pull request merged**